File: services/enrichment/enrichment.service.ts
import 'server-only';

import * as cheerio from 'cheerio';

import type { BusinessContact, BusinessStatus } from '@/types/business';
import type { AppSettings } from '@/types/settings';
import { sleep, withRetry } from '@/utils/async';
import { normalizeHost, normalizePhone } from '@/utils/normalize';
import {
  buildCrawlerUserAgent,
  createHttpClient,
  describeHttpError,
  isRetryableError,
} from '../http/http-client';
import {
  canonicalWhatsApp,
  extractEmails,
  extractPhones,
  extractSocialLinks,
  extractStructuredContacts,
  findContactFormUrl,
  findContactPageLinks,
  type SocialLinks,
} from './extractors';
import { getRobotsChecker } from './robots';
import { checkCrawlTarget } from './url-guard';

const MAX_HTML_BYTES = 3 * 1024 * 1024;

export interface EnrichmentResult {
  contact: BusinessContact;
  status: BusinessStatus;
  pagesVisited: number;
  pagesBlocked: number;
  elapsedMs: number;
  error: string | null;
  notes: string;
}

function emptyContact(): BusinessContact {
  return {
    email: null,
    additionalEmails: [],
    whatsapp: null,
    facebook: null,
    instagram: null,
    linkedin: null,
    twitter: null,
    youtube: null,
    contactFormUrl: null,
    websitePhones: [],
  };
}

function emptyResult(status: BusinessStatus, notes: string, error: string | null = null): EnrichmentResult {
  return {
    contact: emptyContact(),
    status,
    pagesVisited: 0,
    pagesBlocked: 0,
    elapsedMs: 0,
    error,
    notes,
  };
}

interface PageHarvest {
  emails: string[];
  phones: string[];
  social: SocialLinks;
  contactFormUrl: string | null;
  contactLinks: string[];
}

/**
 * Discovers publicly listed contact details from a business website.
 *
 * Scope is deliberately narrow: it fetches the homepage and a small number of
 * linked contact/about pages over plain HTTP, reads what the site publishes, and
 * stops. It honours robots.txt, sends an identifying User-Agent, paces requests,
 * and never attempts to authenticate, submit forms, or work around any access
 * control. Pages behind a login are simply not read.
 */
class EnrichmentService {
  async enrich(
    website: string,
    settings: AppSettings,
    signal: AbortSignal,
  ): Promise<EnrichmentResult> {
    const startedAt = Date.now();

    if (!website) return emptyResult('no-website', 'No website listed');

    let origin: string;
    let homepage: string;
    try {
      const parsed = new URL(website);
      origin = parsed.origin;
      homepage = parsed.toString();
    } catch {
      return emptyResult('enrichment-failed', 'Invalid website URL', 'Invalid URL');
    }

    const targetCheck = await checkCrawlTarget(homepage);
    if (!targetCheck.safe) {
      const reason = targetCheck.reason ?? 'Website host is not reachable';
      return emptyResult('enrichment-failed', reason, reason);
    }

    const userAgent = buildCrawlerUserAgent(settings.crawlerContactEmail);
    const client = createHttpClient({
      timeoutMs: settings.requestTimeoutMs,
      userAgent,
      maxContentLength: MAX_HTML_BYTES,
      maxRedirects: 4,
    });

    const robots = settings.respectRobotsTxt
      ? await getRobotsChecker(origin, userAgent, settings.requestTimeoutMs, signal)
      : { isAllowed: () => true, crawlDelayMs: null };

    // A site asking for a slower crawl gets it, even if our own delay is lower.
    const delayMs = Math.max(settings.delayMs, robots.crawlDelayMs ?? 0);

    const contact = emptyContact();
    const emails: string[] = [];
    const phones = new Set<string>();
    const social: SocialLinks = {};

    let pagesVisited = 0;
    let pagesBlocked = 0;
    let firstError: string | null = null;
    const notes: string[] = [];

    const visit = async (url: string): Promise<PageHarvest | null> => {
      signal.throwIfAborted();

      if (!robots.isAllowed(url)) {
        pagesBlocked += 1;
        return null;
      }

      try {
        const response = await withRetry(
          async () => client.get<string>(url, {
            signal,
            responseType: 'text',
            transformResponse: [(data: unknown) => data],
          }),
          {
            attempts: settings.retryAttempts,
            baseDelayMs: 600,
            signal,
            shouldRetry: isRetryableError,
          },
        );

        // 401/403 means the page is not public; respect that and move on.
        if (response.status === 401 || response.status === 403) {
          pagesBlocked += 1;
          return null;
        }
        if (response.status >= 400) {
          firstError ??= `HTTP ${response.status}`;
          return null;
        }

        const contentType = String(response.headers['content-type'] ?? '');
        if (contentType && !contentType.includes('html')) return null;
        if (typeof response.data !== 'string') return null;

        pagesVisited += 1;
        const $ = cheerio.load(response.data);
        const structured = extractStructuredContacts($);

        return {
          emails: [...structured.emails, ...extractEmails($, url)],
          phones: [...structured.phones, ...extractPhones($)],
          social: extractSocialLinks($, url),
          contactFormUrl: findContactFormUrl($, url),
          contactLinks: findContactPageLinks($, url, settings.maxPagesPerSite),
        };
      } catch (error) {
        if (error instanceof DOMException && error.name === 'AbortError') throw error;
        firstError ??= describeHttpError(error);
        return null;
      }
    };

    const absorb = (harvest: PageHarvest): void => {
      for (const email of harvest.emails) {
        if (!emails.includes(email)) emails.push(email);
      }
      for (const phone of harvest.phones) phones.add(phone);
      for (const [key, value] of Object.entries(harvest.social) as [keyof SocialLinks, string][]) {
        social[key] ??= value;
      }
      contact.contactFormUrl ??= harvest.contactFormUrl;
    };

    const home = await visit(homepage);
    if (home) absorb(home);

    if (!home && pagesBlocked > 0) {
      return {
        ...emptyResult('enrichment-failed', 'Blocked by robots.txt', firstError),
        pagesBlocked,
        elapsedMs: Date.now() - startedAt,
      };
    }

    // Follow a few contact/about pages — that's where addresses usually live.
    const budget = Math.max(0, settings.maxPagesPerSite - 1);
    const candidates = (home?.contactLinks ?? []).slice(0, budget);

    for (const candidate of candidates) {
      if (delayMs > 0) await sleep(delayMs, signal);
      const harvest = await visit(candidate);
      if (harvest) absorb(harvest);
    }

    if (pagesBlocked > 0) notes.push(`${pagesBlocked} page(s) not publicly accessible`);

    const siteHost = normalizeHost(homepage);
    const [primaryEmail, ...rest] = emails;

    contact.email = primaryEmail ?? null;
    contact.additionalEmails = rest.slice(0, 5);
    contact.websitePhones = [...phones].map(normalizePhone).slice(0, 5);
    contact.facebook = social.facebook ?? null;
    contact.instagram = social.instagram ?? null;
    contact.linkedin = social.linkedin ?? null;
    contact.twitter = social.twitter ?? null;
    contact.youtube = social.youtube ?? null;
    contact.whatsapp = social.whatsapp ? canonicalWhatsApp(social.whatsapp) : null;

    const foundAnything =
      Boolean(contact.email) ||
      contact.websitePhones.length > 0 ||
      Object.keys(social).length > 0 ||
      Boolean(contact.contactFormUrl);

    let status: BusinessStatus;
    if (pagesVisited === 0) status = 'enrichment-failed';
    else if (contact.email) status = 'enriched';
    else if (foundAnything) status = 'partial';
    else status = 'partial';

    if (pagesVisited === 0 && !firstError) firstError = 'No readable pages';
    if (siteHost) notes.push(`Crawled ${pagesVisited} page(s) on ${siteHost}`);

    return {
      contact,
      status,
      pagesVisited,
      pagesBlocked,
      elapsedMs: Date.now() - startedAt,
      error: pagesVisited === 0 ? firstError : null,
      notes: notes.join('; '),
    };
  }
}

export const enrichmentService = new EnrichmentService();

File: backend/src/LeadMine.Infrastructure/Scraping/Enrichment/EnrichmentService.cs
using System.Diagnostics;
using System.Net;
using System.Text;
using HtmlAgilityPack;
using LeadMine.Domain.Enums;
using Microsoft.Extensions.Logging;

namespace LeadMine.Infrastructure.Scraping.Enrichment;

/// <summary>Contact details discovered on a business website.</summary>
public sealed class BusinessContact
{
    public string? Email { get; set; }
    public List<string> AdditionalEmails { get; set; } = [];
    public string? WhatsApp { get; set; }
    public string? Facebook { get; set; }
    public string? Instagram { get; set; }
    public string? LinkedIn { get; set; }
    public string? Twitter { get; set; }
    public string? YouTube { get; set; }
    public string? ContactFormUrl { get; set; }
    public List<string> WebsitePhones { get; set; } = [];
}

public sealed class EnrichmentResult
{
    public required BusinessContact Contact { get; init; }
    public required BusinessStatus Status { get; init; }
    public int PagesVisited { get; init; }
    public int PagesBlocked { get; init; }
    public long ElapsedMs { get; init; }
    public string? Error { get; init; }
    public string Notes { get; init; } = string.Empty;

    public static EnrichmentResult Empty(BusinessStatus status, string notes, string? error = null) => new()
    {
        Contact = new BusinessContact(),
        Status = status,
        Notes = notes,
        Error = error,
    };
}

/// <summary>
/// Discovers publicly listed contact details from a business website.
/// <para>
/// Scope is deliberately narrow: it fetches the homepage and a small number of
/// linked contact/about pages over plain HTTP, reads what the site publishes,
/// and stops. It honours robots.txt, sends an identifying User-Agent, paces
/// requests, and never attempts to authenticate, submit a form, or work around
/// any access control. Pages behind a login are simply not read.
/// </para>
/// </summary>
public sealed class EnrichmentService(
    IHttpClientFactory httpClientFactory,
    RobotsCache robotsCache,
    ILogger<EnrichmentService> logger)
{
    private const int MaxHtmlBytes = 3 * 1024 * 1024;

    public async Task<EnrichmentResult> EnrichAsync(
        string website,
        ScraperOptions options,
        CancellationToken ct = default)
    {
        var stopwatch = Stopwatch.StartNew();

        if (string.IsNullOrWhiteSpace(website))
        {
            return EnrichmentResult.Empty(BusinessStatus.NoWebsite, "No website listed");
        }

        if (!Uri.TryCreate(website, UriKind.Absolute, out var homepage) ||
            (homepage.Scheme != Uri.UriSchemeHttp && homepage.Scheme != Uri.UriSchemeHttps))
        {
            return EnrichmentResult.Empty(BusinessStatus.EnrichmentFailed, "Invalid website URL", "Invalid URL");
        }

        var targetCheck = await UrlGuard.CheckAsync(homepage.ToString(), ct);

        if (!targetCheck.Safe)
        {
            var reason = targetCheck.Reason ?? "Website host is not reachable";
            return EnrichmentResult.Empty(BusinessStatus.EnrichmentFailed, reason, reason);
        }

        var origin = new Uri(homepage.GetLeftPart(UriPartial.Authority));

        var robots = options.RespectRobotsTxt
            ? await robotsCache.GetAsync(origin, options.UserAgent, ct)
            : RobotsPolicy.Allow;

        // A site asking for a slower crawl gets it, even if our own delay is lower.
        var delay = TimeSpan.FromMilliseconds(
            Math.Max(options.DelayMs, robots.CrawlDelay?.TotalMilliseconds ?? 0));

        var state = new CrawlState();
        var home = await VisitAsync(homepage, robots, options, state, ct);

        if (home is null && state.PagesBlocked > 0)
        {
            return new EnrichmentResult
            {
                Contact = new BusinessContact(),
                Status = BusinessStatus.EnrichmentFailed,
                Notes = "Blocked by robots.txt",
                Error = state.FirstError ?? "Blocked by robots.txt",
                PagesBlocked = state.PagesBlocked,
                ElapsedMs = stopwatch.ElapsedMilliseconds,
            };
        }

        if (home is not null) Absorb(state, home);

        // Follow a few contact/about pages — that's where addresses usually live.
        var budget = Math.Max(0, options.MaxPagesPerSite - 1);

        foreach (var candidate in (home?.ContactLinks ?? []).Take(budget))
        {
            if (delay > TimeSpan.Zero) await Task.Delay(delay, ct);

            if (!Uri.TryCreate(candidate, UriKind.Absolute, out var candidateUri)) continue;

            var harvest = await VisitAsync(candidateUri, robots, options, state, ct);
            if (harvest is not null) Absorb(state, harvest);
        }

        return Finish(state, homepage, stopwatch);
    }

    private sealed class CrawlState
    {
        public List<string> Emails { get; } = [];
        public List<string> Phones { get; } = [];
        public Dictionary<SocialKey, string> Social { get; } = [];
        public string? ContactFormUrl { get; set; }
        public int PagesVisited { get; set; }
        public int PagesBlocked { get; set; }
        public string? FirstError { get; set; }
    }

    private async Task<PageHarvest?> VisitAsync(
        Uri url,
        RobotsPolicy robots,
        ScraperOptions options,
        CrawlState state,
        CancellationToken ct)
    {
        ct.ThrowIfCancellationRequested();

        if (!robots.IsAllowed(url))
        {
            state.PagesBlocked++;
            return null;
        }

        try
        {
            var client = httpClientFactory.CreateClient(ScraperHttpClients.Crawler);

            using var response = await ScraperRetry.ExecuteAsync(
                async _ =>
                {
                    using var request = new HttpRequestMessage(HttpMethod.Get, url);
                    var result = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);

                    if (ScraperRetry.IsRetryableStatus(result.StatusCode))
                    {
                        var status = result.StatusCode;
                        result.Dispose();
                        throw new HttpRequestException($"HTTP {(int)status}", null, status);
                    }

                    return result;
                },
                options.RetryAttempts,
                logger,
                ct);

            // 401/403 means the page is not public; respect that and move on.
            if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                state.PagesBlocked++;
                return null;
            }

            if (!response.IsSuccessStatusCode)
            {
                state.FirstError ??= $"HTTP {(int)response.StatusCode}";
                return null;
            }

            var contentType = response.Content.Headers.ContentType?.MediaType ?? string.Empty;
            if (contentType.Length > 0 && !contentType.Contains("html", StringComparison.OrdinalIgnoreCase))
            {
                return null;
            }

            var html = await RobotsCache.ReadCappedAsync(response, MaxHtmlBytes, ct);
            if (html.Length == 0) return null;

            state.PagesVisited++;

            var document = new HtmlDocument();
            document.LoadHtml(html);

            // The response URL, not the requested one: redirects change what
            // "same host" and relative links mean.
            var effectiveUrl = response.RequestMessage?.RequestUri ?? url;

            return ContactExtractor.Harvest(document, effectiveUrl, options.MaxPagesPerSite);
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception ex)
        {
            state.FirstError ??= ex.Message;
            logger.LogDebug(ex, "Could not read {Url}", url);
            return null;
        }
    }

    private static void Absorb(CrawlState state, PageHarvest harvest)
    {
        foreach (var email in harvest.Emails)
        {
            if (!state.Emails.Contains(email, StringComparer.OrdinalIgnoreCase)) state.Emails.Add(email);
        }

        foreach (var phone in harvest.Phones)
        {
            if (!state.Phones.Contains(phone, StringComparer.OrdinalIgnoreCase)) state.Phones.Add(phone);
        }

        foreach (var (key, value) in harvest.Social) state.Social.TryAdd(key, value);

        state.ContactFormUrl ??= harvest.ContactFormUrl;
    }

    private static EnrichmentResult Finish(CrawlState state, Uri homepage, Stopwatch stopwatch)
    {
        var contact = new BusinessContact
        {
            Email = state.Emails.FirstOrDefault(),
            AdditionalEmails = state.Emails.Skip(1).Take(5).ToList(),
            WebsitePhones = state.Phones.Take(5).ToList(),
            ContactFormUrl = state.ContactFormUrl,
            Facebook = state.Social.GetValueOrDefault(SocialKey.Facebook),
            Instagram = state.Social.GetValueOrDefault(SocialKey.Instagram),
            LinkedIn = state.Social.GetValueOrDefault(SocialKey.LinkedIn),
            Twitter = state.Social.GetValueOrDefault(SocialKey.Twitter),
            YouTube = state.Social.GetValueOrDefault(SocialKey.YouTube),
        };

        if (state.Social.TryGetValue(SocialKey.WhatsApp, out var whatsapp))
        {
            contact.WhatsApp = ContactExtractor.CanonicalWhatsApp(whatsapp);
        }

        var status = state.PagesVisited == 0
            ? BusinessStatus.EnrichmentFailed
            : contact.Email is not null
                ? BusinessStatus.Enriched
                : BusinessStatus.Partial;

        var notes = new StringBuilder();

        if (state.PagesBlocked > 0) notes.Append($"{state.PagesBlocked} page(s) not publicly accessible");

        var siteHost = ContactExtractor.NormalizeHost(homepage.ToString());

        if (siteHost.Length > 0)
        {
            if (notes.Length > 0) notes.Append("; ");
            notes.Append($"Crawled {state.PagesVisited} page(s) on {siteHost}");
        }

        return new EnrichmentResult
        {
            Contact = contact,
            Status = status,
            PagesVisited = state.PagesVisited,
            PagesBlocked = state.PagesBlocked,
            ElapsedMs = stopwatch.ElapsedMilliseconds,
            // An error is only worth surfacing when nothing was read at all —
            // one broken sub-page on an otherwise-crawled site is not a failure.
            Error = state.PagesVisited == 0 ? state.FirstError ?? "No readable pages" : null,
            Notes = notes.ToString(),
        };
    }
}
